// eligo/query-engine/combinators.js

var utils = require('protean');

function combinatorToString () {
    return this.key;
}
//---------------------------------------------------------------------------
// Exports
//---------------------------------------------------------------------------
Object.defineProperties(exports, {
    define: { value: function define (key, token, fn) {
        this._tokens = null;
        
        this[key] = this[token.source] = this[fn.name] = fn;
        fn.key = key;
        fn.token = token;
        fn.type = 'combinator';
        fn.toString = combinatorToString;
        
        return fn;
    }, enumerable: true },
    
    tokens: { get: function () {
        if (!this._tokens) {
            this._tokens = utils.chain(this).
                keys().
                map(function (key) {
                    var fn;
                    if (key !== 'define' &&
                        key !== 'tokens' &&
                        key !== 'parse' &&
                        (fn = exports[key]) &&
                        typeof fn === 'function' &&
                        fn.token instanceof RegExp
                    ) {
                        return fn.token;
                    }
                }).
                compact().
                sort(function (a, b) {
                    return b.source.length - a.source.length;
                }).
                value();
        }
        return this._tokens;
    }, enumerable: true },
    
    parse: { value: function (scanner) {
        var combinator = null;
        
        this.tokens.
            some(function (token) {
                if (scanner.scan(token)) {
                    combinator = exports[token.source];
                    return true;
                }
            });
        
        return combinator;
    }, enumerable: true }
});
//---------------------------------------------------------------------------
// Combinators
//---------------------------------------------------------------------------
exports.define(' ', (/\s*/), function descendent (path, test) {
    var p = path.parent;
    
    while (p && !test(p)) {
        p = path.parent;
    }
    
    return p ? true : false;
});

exports.define('>', (/\s*>\s*/), function child (path, test) {
    var node = path.node,
        p = path.parent,
        children = path.getChildren();
    
    return test(p) && ~children.indexOf(node);
});

exports.define('~', (/\s*~\s*/), function sibling (path, test) {
    var node = path.node,
        p = path.parent,
        children = path.getChildren(),
        i = children.indexOf(node);
    
    if (i > 0) {
        while (i--) {
            if (test(children[i])) {
                return true;
            }
        }
    }
    
    return false;
});

exports.define('+', (/\s*\+\s*/), function adjacentSibling (path, test) {
    var node = path.node,
        p = path.parent,
        children = path.getChildren(),
        nodeIdx = children.indexOf(node);
    
    return ~nodeIdx && test(children[nodeIdx - 1]);
});
