// eligo/query-engine/scanner.js

var utils = require('protean'),
    PStrScan = require('pstrscan'),
    Scanner;

Scanner = utils.inherit(PStrScan, {
    constructor: function (opts) {
        var aliases;

        opts = opts || {};

        Scanner.superclass.call(this);

        this.typeKey  = opts.typeKey;
        this.idKey    = opts.idKey;

        aliases = opts.aliases;

        if (aliases) {
            this.aliases  = aliases;

            this.aliasesRegExp = new RegExp('(' +
                utils.keys(aliases).
                    sort(function (a, b) {
                        return a.length - b.length;
                    }).
                    map(function (key) {
                        return aliases[key].source;
                    }).
                    join('|') +
                ')');
        }
    },

    replace: function (text) {
        var aliases = this.aliases;
        
        return aliases ?
            text.replace(this.aliasesRegExp, function (match) {
                return aliases[match];
            })
            :
            text;
    }
});

//---------------------------------------------------------------------------
// Exports
//---------------------------------------------------------------------------
module.exports = Scanner;
