// eligo/query-engine/tokens.js

var utils = require('eligo/utils'),
    combinators = require('./combinators'),
    operators = require('./operators');

function regExpSource (re) {
    return re.source;
}

function regExpUnion (list) {
    return new RegExp('(' + list.map(regExpSource).join('|') + ')');
}

utils.extend(exports, {
    UNIVERSAL:          /\*/,
    GROUP_SEPARATOR:    /\s*,\s*/,
    WHITESPACE:         /\s*/,
    IDENTIFIER:         /[A-Za-z_$][A-Za-z0-9_$\-]+/,
    ATTRIBUTE_START:    /\[\s*/,
    ATTRIBUTE_END:      /\s*\]/,
    QUOTES:             /["']/,
});

Object.defineProperties(exports, {
    TYPE: {
        value: new RegExp('(' + exports.IDENTIFIER.source + ')'),
        enumerable: true
    },
    ID: {
        value: new RegExp('#(' + exports.IDENTIFIER.source + ')'),
        enumerable: true
    },
    MEMBER: {
        value: new RegExp('\\.(' + exports.IDENTIFIER.source + ')'),
        enumerable: true
    },
    COMBINATOR: {
        get: function () {
            return regExpUnion(combinators.tokens);
        },
        enumerable: true
    },
    OPERATOR: {
        get: function () {
            return regExpUnion(operators.tokens);
        },
        enumerable: true
    }
});
